src/storage/src/source/reclock.rs

```diff
@@ -340,50 +339,56 @@ where
     }
 
     pub fn compact(&mut self, new_since: Antichain<IntoTime>) {
-        let mut inner = &mut *self.inner.borrow_mut();
         // Ignore compaction requests while we initialize
-        if !PartialOrder::less_than(&inner.as_of, &inner.upper) {
+        if !self.initialized() {
             return;
         }
-        if !PartialOrder::less_equal(&inner.since, &new_since) {
+        let inner = &mut *self.inner.borrow_mut();
+        if !PartialOrder::less_equal(&self.since, &new_since) {
             panic!(
                 "ReclockFollower: `new_since` ({:?}) is not beyond \
                 `self.since` ({:?}).",
-                new_since, inner.since,
+                new_since, self.since,
             );
         }
+        inner.since.update_iter(
+            self.since
+                .iter()
+                .map(|t| (t.clone(), -1))
+                .chain(new_since.iter().map(|t| (t.clone(), 1))),
+        );
```

Comment:
`MutableAntichain` makes this :chef_kiss: to have multiple people concurrently alter this!

But I have one question: if one handle to the follower calls compact and the other doesn't (which I believe is what happens in the pipeline), wont the latter have a local `since` field that is quite old?

Reply:
> wont the latter have a local since field that is quite old?

The handle that doesn't would cause compaction to never happen, yes

> which I believe is what happens in the pipeline

It was but I changed it. See the change and comment in `source_reader_pipeline.rs`